File: java/src/edu/caltech/ipac/firefly/core/background/PackageProgress.java
package edu.caltech.ipac.firefly.core.background;
/**
 * User: roby
 * Date: 6/6/14
 * Time: 11:22 AM
 */


import edu.caltech.ipac.util.HandSerialize;
import edu.caltech.ipac.util.StringUtils;

import java.io.Serializable;

/**
* @author Trey Roby
*/
public class PackageProgress implements Serializable, HandSerialize {
    private final static String SPLIT_TOKEN= "--BGProgress--";
    private int _totalFiles= 0;
    private int _processedFiles= 0;
    private long _totalBytes= 0;
    private long _processedBytes= 0;
    private long _finalCompressedBytes= 0;
    private String _url= null;

    PackageProgress() {}


    public PackageProgress(int totalFiles,
                           int processedFiles,
                           long totalBytes,
                           long processedBytes,
                           long finalCompressedBytes,
                           String url) {
        _totalFiles = totalFiles;
        _processedFiles = processedFiles;
        _totalBytes = totalBytes;
        _processedBytes = processedBytes;
        _finalCompressedBytes = finalCompressedBytes;
        _url= url;
    }

    public int getTotalFiles() {
        return _totalFiles;
    }

    public int getProcessedFiles() {
        return _processedFiles;
    }

    public long getTotalByes() {
        return _totalBytes;
    }

    public long getProcessedBytes() {
        return _processedBytes;
    }

    public long getFinalCompressedBytes() {
        return _finalCompressedBytes;
    }

    public String getURL() { return _url; }

    public boolean isDone() { return _url!=null; }

    public String serialize() {
        return StringUtils.combine(SPLIT_TOKEN,
                                   _totalFiles + "",
                                   _processedFiles + "",
                                   _totalBytes + "",
                                   _processedBytes + "",
                                   _finalCompressedBytes + "",
                                   _url);
    }
    public static PackageProgress parse(String s) {
        if (s==null) return null;
        PackageProgress p= new PackageProgress();
        try {
            String sAry[]= StringUtils.parseHelper(s,6,SPLIT_TOKEN);
            int i= 0;
            p._totalFiles= StringUtils.getInt(sAry[i++],0);
            p._processedFiles= StringUtils.getInt(sAry[i++],0);
            p._totalBytes= StringUtils.getLong(sAry[i++],0);
            p._processedBytes= StringUtils.getLong(sAry[i++],0);
            p._finalCompressedBytes= StringUtils.getLong(sAry[i++],0);
            p._url= StringUtils.checkNull(sAry[i++]);
        } catch (IllegalArgumentException e) {
            p= null;
        }
        return p;
    }
}

File: src/firefly/js/ipac-firefly/core/background/PackageProgress.js
/*
 * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
 */

/*jshint browserify:true*/
/*jshint esnext:true*/
/*jshint curly:false*/
/**
 * User: roby
 * Date: Apr 2, 2009
 * Time: 9:18:47 AM
 */
"use strict";
import validator from "validator";


const SPLIT_TOKEN= "--BGProgress--";

class PackageProgress {
    /**
     *
     * @param {number} totalFiles
     * @param {number} processedFiles
     * @param {number} totalBytes
     * @param {number} processedBytes
     * @param {number} finalCompressedBytes
     * @param {string} url
     */
    construtor(totalFiles=0, processedFiles=0, totalBytes=0, processedBytes=0, finalCompressedBytes=0, url=null) {
        this.totalFiles = totalFiles;
        this.processedFiles = processedFiles;
        this.totalBytes = totalBytes;
        this.processedBytes = processedBytes;
        this.finalCompressedBytes = finalCompressedBytes;
        this.url= url;
    }

    getTotalFiles() { return this.totalFiles; }

    getProcessedFiles() { return this.processedFiles; }

    getTotalByes() { return this.totalBytes; }

    getProcessedBytes() { return this.processedBytes; }

    getFinalCompressedBytes() { return this.finalCompressedBytes; }

    getURL() { return this.url; }

    /**
     * if a url exist then we are done
     * @return {boolean} true if done
     */
    isDone() { return this.url?true:false; }

    serialize() {
        var retval= [_totalFiles,
                     _processedFiles,
                     _totalBytes,
                     _processedBytes,
                     _finalCompressedBytes,
                     _url].join(SPLIT_TOKEN);

        return retval;
    }

    /**
     * parse a serialized version into a PackageProgress
     * @param s
     * @return {PackageProgress}
     */
    static parse(s) {
        if (!s) return null;
        var p= null;
        var sAry= s.split(SPLIT_TOKEN,6);
        if (sAry.length===6) {
            p= new PackageProgress();
            p._totalFiles= validator.toInt(sAry[i++],0);
            p._processedFiles= validator.toInt(sAry[i++],0);
            p._totalBytes= validator.toInt(sAry[i++],0);
            p._processedBytes= validator.toInt(sAry[i++],0);
            p._finalCompressedBytes= validator.toInt(sAry[i++],0);
            if (p._totalFiles && p._processedFiles && p._totalBytes && p._processedBytes && p._finalCompressedBytes) {
                p._url= sAry[i++];
            }
            else {
                p=null;
            }
        }
        return p;
    }
}

